readthedocs/api/v2/views/model_views.py

```diff
@@ -300,14 +325,30 @@ def reset(self, request, **kwargs):
         instance.reset()
         return Response(status=status.HTTP_204_NO_CONTENT)
 
+    def get_queryset_for_api_key(self, api_key):
+        return self.model.objects.filter(project=api_key.project)
+
 
 class BuildCommandViewSet(DisableListEndpoint, CreateModelMixin, UserSelectViewSet):
     parser_classes = [JSONParser, MultiPartParser]
-    permission_classes = [APIRestrictedPermission]
+    permission_classes = [HasBuildAPIKey | APIRestrictedPermission]
     renderer_classes = (JSONRenderer,)
     serializer_class = BuildCommandSerializer
     model = BuildCommandResult
 
+    def perform_create(self, serializer):
+        """Restrict creation to builds attached to the project from the api key."""
+        build_pk = serializer.validated_data["build"].pk
+        api_key = self.request.build_api_key
+        if api_key and not api_key.project.builds.filter(pk=build_pk).exists():
+            raise PermissionDenied()
+        # If the request isn't attached to a build api key,
+        # the user doing the request is a superuser, so it has access to all projects.
+        return super().perform_create(serializer)
```

Comment:
Why do we need this only for this view but it's not required when updating a Version or Build object via the API from the builder as well?

Reply:
When updating, we know the object (since it already exists), so the queryset already filters the objects the token has access to. When creating, we don't know where the object belongs since it hasn't been created, but we know that it should be attached to the project the token grants access to.